src/main/java/com/jwplayer/southpaw/util/ByteArraySet.java

```diff
@@ -219,59 +222,104 @@ public static Chunk deserializeCompact(byte[] bytes, int from, int to) {
+                    return true;
+                }
+                index += entrySize + 1;
+            }
+            return false;
+        }
+
+        /**
+         * Removes the given ByteArrays from this chunk by filling in its section of the chunk with 0s.
+         * @param valsToRemove - The ByteArrays to remove
+         * @return True if any value was removed, otherwise false
+         */
+        public boolean removeAll(Set<ByteArray> valsToRemove) {
+            if(entries == 0) return false;
+
+            int index = 0;
+            boolean valsRemoved = false;
+            for(ByteArray valToRemove: valsToRemove) {
+                if(valToRemove != null && valToRemove.size() > 0 &&
+                        min.compareTo(valToRemove) <= 0 && max.compareTo(valToRemove) >= 0) {
+                    while(index < size) {
+                        int entrySize = bytes[index];
+                        if(entrySize == 0) {
+                            // This is a previously deleted value, so move the index forward
+                            index++;
+                            continue;
                         }
-                        entries--;
-                        if(entries > 0) {
-                            if(min.equals(byteArray)) {
-                                index = 0;
-                                while(index < size) {
-                                    baSize = (int) bytes[index];
-                                    index++;
-                                    if(baSize != 0) {
-                                        min = new ByteArray(Arrays.copyOfRange(bytes, index, index + baSize));
-                                        break;
-                                    }
-                                    index += baSize;
-                                }
-                            }
-                            if(max.equals(byteArray)) {
-                                index = 0;
-                                while(index < size) {
-                                    baSize = (int) bytes[index];
-                                    index++;
-                                    if(baSize != 0) {
-                                        max = new ByteArray(Arrays.copyOfRange(bytes, index, index + baSize));
-                                    }
-                                    index += baSize;
-                                }
-                            }
+                        int comparison = valToRemove.compareTo(bytes, index + 1, entrySize);
+                        if(comparison < 0) {
+                            // valToRemove is less than the current index, so move to the next valToRemove
+                            break;
+                        } else if(comparison == 0) {
+                            Arrays.fill(bytes, index, index + entrySize + 1, (byte) 0);
+                            entries--;
+                            index += entrySize + 1;
```

Comment:
Shouldn't you start traversing `bytes` from the beginning when you change you move to the next `valToRemove`?

Reply:
index starts at 0. If valToRemove is less than the value at the current index, it goes to the next valToRemove. If valToRemove is larger, it progresses the index. If they are equal, it zeroes out the value at the current index. If we're at the end, all possible values have been removed.

Reply:
It assumes that `valsToRemove` is sorted then, right?